server/src/main/java/org/elasticsearch/common/time/DateUtils.java

```diff
@@ -77,4 +78,224 @@ public static ZoneId of(String zoneId) {
+     *
+     * Note: This does not check for correctness of the result, as this only works with units smaller or equal than a day
+     *       In order to ensure the performane of this methods, there are no guards or checks in it
+     *
+     * @param utcMillis   the milliseconds since the epoch
+     * @param unitMillis  the unit to round to
+     * @return            the rounded milliseconds since the epoch
+     */
+    public static long roundFloor(long utcMillis, long unitMillis) {
+        if (utcMillis >= 0) {
+            return utcMillis - utcMillis % unitMillis;
+        } else {
+            utcMillis += 1;
+            return utcMillis - utcMillis % unitMillis - unitMillis;
+        }
+    }
+
+    /**
+     * Round down to the beginning of the quarter of the year of the specified time
+     * @param utcMillis the milliseconds since the epoch
+     * @return The milliseconds since the epoch rounded down to the quarter of the year
+     */
+    public static long roundQuarterOfYear(long utcMillis) {
+        int year = DateUtils.getYear(utcMillis);
+        int month = DateUtils.getMonthOfYear(utcMillis, year);
+        return DateUtils.of(year, Month.of(month).firstMonthOfQuarter().getValue());
+    }
+
+    /**
+     * Round down to the beginning of the month of the year of the specified time
+     * @param utcMillis the milliseconds since the epoch
+     * @return The milliseconds since the epoch rounded down to the month of the year
+     */
+    public static long roundMonthOfYear(long utcMillis) {
+        int year = DateUtils.getYear(utcMillis);
+        int month = DateUtils.getMonthOfYear(utcMillis, year);
+        return DateUtils.of(year, month);
+    }
+
+    /**
+     * Round down to the beginning of the beginning of the year of the specified time
+     * @param utcMillis the milliseconds since the epoch
+     * @return The milliseconds since the epoch rounded down to the beginning of the year
+     */
+    public static long getFirstDayOfYearMillis(long utcMillis) {
+        int year = getYear(utcMillis);
+        return calculateFirstDayOfYearMillis(year);
+    }
+
+    /**
+     * Return the first day of the month
+     * @param year  the year to return
+     * @param month the month to return, ranging from 1-12
+     * @return the milliseconds since the epoch of the first day of the month in the year
+     */
+    private static long of(int year, int month) {
+        long millis = getYearMillis(year);
+        millis += getTotalMillisByYearMonth(year, month);
+        return millis;
+    }
+
+    /**
+     * calculates the first day of a year in milliseconds since the epoch (assuming UTC)
+     *
+     * @param year the year
+     * @return the milliseconds since the epoch of the first of january at midnight of the specified year
+     */
+    // see org.joda.time.chrono.GregorianChronology
+    private static long calculateFirstDayOfYearMillis(int year) {
```

Comment:
I find the mention of `Day` in this method's name to be a little confusing. Maybe `utcMillisAtStartOfYear`?